src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs

```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
 
             public int[] ToArray()
             {
-                int[] array = new int[_end - _start];
-                Fill(array, _start);
+                int start = _start;
+                int[] array = new int[_end - start];
+                Fill(array, start);
                 return array;
             }
 
             public List<int> ToList()
             {
-                List<int> list = new List<int>(_end - _start);
-                Fill(SetCountAndGetSpan(list, _end - _start), _start);
+                int start = _start;
+                int end = _end;
+                List<int> list = new List<int>(end - start);
+                Fill(SetCountAndGetSpan(list, end - start), start);
                 return list;
             }
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void Fill(Span<int> destination, int start)
             {
-                for (int i = 0; i < destination.Length; i++, value++)
+                // The improvements in the compiler now allow us to write the loop as Vector256.
+                // Manually unrolling to Vector128x2 for SSE2/NEON is no longer profitable.
+                if (destination.Length < Vector256<int>.Count)
                 {
-                    destination[i] = value;
+                    goto Scalar;
+                }
+
+                int stride = Vector256<int>.Count;
```

Comment:
Any reason to use V256 instead of V128 to support both x86 and arm?

Reply:
Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?

Reply:
Also, I don't see any guards for SIMD at all - arm32, linux-x86, mono-interp, HWINTRINSIC=0

Reply:
> Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?
Let me check. Also, is it really as low as SSE2? That would be unfortunate.

Reply:
> > Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?
> 
> Let me check. Also, is it really as low as SSE2? That would be unfortunate.

Yes, with SSE4.1 as opportunistic check as far as I remember (runtime-check)

Reply:
`DOTNET_EnableSSE41=0`
```asm
; 2 inlinees with PGO data; 29 single block inlinees; 11 inlinees without PGO data

G_M000_IG01:                ;; offset=0000H
       push     rsi
       sub      rsp, 64

G_M000_IG02:                ;; offset=0005H
       mov      esi, dword ptr [rcx+08H]
       mov      edx, dword ptr [rcx+0CH]
       sub      edx, esi
       movsxd   rdx, edx
       mov      rcx, 0x7FFD85693060
       call     CORINFO_HELP_NEWARR_1_VC
       lea      rdx, bword ptr [rax+10H]
       mov      ecx, dword ptr [rax+08H]
       cmp      ecx, 8
       jl       SHORT G_M000_IG06

G_M000_IG03:                ;; offset=002BH
       movd     xmm0, esi
       pshufd   xmm0, xmm0, 0
       movups   xmm1, xmmword ptr [reloc @RWD00]
       paddd    xmm1, xmm0
       movups   xmm2, xmmword ptr [reloc @RWD16]
       paddd    xmm0, xmm2
       movups   xmm2, xmmword ptr [reloc @RWD32]
       movups   xmm3, xmmword ptr [reloc @RWD32]
       align    [8 bytes for IG04]

G_M000_IG04:                ;; offset=0060H
       movups   xmmword ptr [rdx], xmm1
       movups   xmmword ptr [rdx+10H], xmm0
       paddd    xmm1, xmm3
       paddd    xmm0, xmm2
       add      rdx, 32
       sub      ecx, 8
       cmp      ecx, 8
       jge      SHORT G_M000_IG04

G_M000_IG05:                ;; offset=007BH
       movaps   xmmword ptr [rsp+20H], xmm1
       movaps   xmmword ptr [rsp+30H], xmm0
       mov      esi, dword ptr [rsp+20H]

G_M000_IG06:                ;; offset=0089H
       xor      r8d, r8d
       test     ecx, ecx
       jle      SHORT G_M000_IG08
       align    [0 bytes for IG07]

G_M000_IG07:                ;; offset=0090H
       lea      r9d, [rsi+01H]
       mov      r10d, r8d
       mov      dword ptr [rdx+4*r10], esi
       inc      r8d
       cmp      r8d, ecx
       mov      esi, r9d
       jl       SHORT G_M000_IG07

G_M000_IG08:                ;; offset=00A6H
       add      rsp, 64
       pop      rsi
       ret

RWD00   dq      0000000100000000h, 0000000300000002h
RWD16   dq      0000000500000004h, 0000000700000006h
RWD32   dq      0000000800000008h, 0000000800000008h

; Total bytes of code 172
```
Very similar to ARM64. I have added check for V128 acceleration too.

Reply:
Do we use this approach anywhere else in BCL? Last time we discussed this, V256 via V128x2 was mainly used for testing only

Reply:
How do you want the implementation to look like? Full span.IndexOf style with V256 -> V128 -> Scalar blocks?

Reply:
> How do you want the implementation to look like?

I am fine with the codegen you posted - I am just not aware if this approach is recommended/used anywhere. cc @tannergooding

Reply:
> How do you want the implementation to look like?

Has something changed since the previous PR?  My feedback there was that without motivating scenarios, this wasn't worth vectorizing at all.  Can you point to places where `Enumerable.Range(...).ToArray/ToList` is being used in production code on perf-sensitive code paths?

Reply:
Thank you for reviewing the change. Here are the concerns listed in the original PR that were stating the impl. was not worth it:
- Negative impact on code size
- Negative impact on maintainability
- Negative impact on un-vectorized path

I tried to address the first two by keeping the implementation as simple as possible and evaluating codegen size diff, which is +60B on x86_64 and +100B on Aarch64. The third one is addressed by getting rid of `length % Vector<int>.Count` which was causing regressions on x86_64.

It is true that the pattern most often used in tests, but I don't agree with the assertion that it is not a useful change. There are also references to non-test code that could benefit:
- https://github.com/fabsenet/adrilight/blob/main/adrilight/ViewModel/SettingsViewModel.cs#L55-L56
- https://github.com/SciSharp/Pandas.NET/blob/master/src/Pandas.NET/Methods/Pandas.read_csv.cs#L43
- https://github.com/teenageengineering/videolab/blob/master/Assets/Skinner/SkinnerModel.cs#L67

Matches: https://grep.app/search?q=Enumerable%5C.Range%5C%28%5B%5E%29%5D%2B%5C%29%5C.%28ToArray%7CToList%29%5C%28%5C%29&regexp=true&case=true&filter[lang][0]=C%23 (22 pages of results)